File: src/main/java/dev/reeve/mazelib/generators/RecursiveBacktrackGenerator.java
package dev.reeve.mazelib.generators;

import dev.reeve.mazelib.Maze;
import dev.reeve.mazelib.MazePoint;
import dev.reeve.mazelib.MazePosition;
import lombok.val;
import lombok.var;

import java.util.Random;
import java.util.Stack;

public class RecursiveBacktrackGenerator extends MazeGenerator {
	private final Random random;
	
	public RecursiveBacktrackGenerator() {
		this(new Random());
	}
	
	public RecursiveBacktrackGenerator(Random random) {
		this.random = random;
	}
	
	@Override
	public Maze generateMaze(int sizeX, int sizeY) {
		var updateOrder = 0;
		val maze = new Maze(sizeX, sizeY);
		var positions = new Stack<MazePosition>();
		var currentLocation = new MazePosition(random.nextInt(sizeX), random.nextInt(sizeY));
		MazePosition next;
		
		positions.push(currentLocation);
		
		while (!positions.isEmpty()) {
			if (positions.indexOf(currentLocation) != positions.lastIndexOf(currentLocation)) {
				return maze;
			}
			
			currentLocation = positions.peek();
			
			val direction = generateDirection(maze, random, currentLocation);
			
			
			if (!maze.isPointSet(currentLocation)) {
				maze.setPoint(currentLocation, new MazePoint(currentLocation, updateOrder));
			}
			
			if (direction == null) {
				positions.pop();
				continue;
			}
			
			maze.setOpen(maze.getPoint(currentLocation), direction);
			
			maze.getPoint(currentLocation).setUpdateOrder(updateOrder);
			
			next = currentLocation.inDirection(direction);
			val newPoint = new MazePoint(next, updateOrder);
			maze.setPoint(next, newPoint);
			// make sure the wall to get back to the last square in the path is open
			maze.setOppositeOpen(newPoint, direction);
			
			positions.push(next.clone());
			
			updateOrder++;
		}
		
		return maze;
	}
}

File: src/main/kotlin/dev/reeve/mazelib/generators/RecursiveBacktrackGenerator.kt
package dev.reeve.mazelib.generators

import dev.reeve.mazelib.*
import dev.reeve.mazelib.solvers.IncompleteMazeException
import java.util.*

/**
 * Generate a maze with a random walk, and backtracking when it finds a dead end.
 * @param mask Passed to [Maze]
 */
class RecursiveBacktrackGenerator(var mask: (MazePosition) -> Boolean = { false }, random: Random) : MazeGeneratorWithSolution(random) {
	override fun generateMaze(sizeX: Int, sizeY: Int): Maze {
		return generate(sizeX, sizeY, MazePosition(random.nextInt(sizeX), random.nextInt(sizeY)), null).first
	}
	
	private fun generate(sizeX: Int, sizeY: Int, start: MazePosition, end: MazePosition?): Pair<Maze, MazePath?> {
		val maze = Maze(sizeX, sizeY)
		val run = MazePath(start)
		var last: MazeDirection? = null
		var order = 0
		var path: MazePath? = null
		
		while (run.isNotEmpty()) {
			if (order > 1000) error("Maze took too many tries")
			
			val currentPos = run.peek()
			val direction = generateDirection(maze, currentPos)
			
			if (end != null && path == null && currentPos == end) {
				path = (run.clone() as MazePath)
			}
			
			val point = maze.getPoint(currentPos)?.also {
				if (!maze.isPointSet(currentPos))
					it.updateOrder = order++
			} ?: MazePoint(position = currentPos, updateOrder = order++)
			
			if (!maze.isPointSet(currentPos)) {
				maze.setPoint(point)
				
				if (last != null) {
					maze.setOppositeOpen(point, last)
				}
			}
			
			if (direction == null) {
				run.pop()
				continue
			}
			
			maze.setOpen(point, direction)
			
			last = direction
			run.add(currentPos.inDirection(direction))
		}
		
		return maze to path
	}
	
	override fun generateMaze(sizeX: Int, sizeY: Int, start: MazePosition, end: MazePosition): Pair<Maze, MazePath> {
		return generate(sizeX, sizeY, start, end).let { it.first to (it.second ?: throw IncompleteMazeException()) }
	}
}